File: src/main/java/minic/uilocal/visual/MiniCSemanticScopeTreeModelFactory.java
package minic.uilocal;

import minic.uiapi.UiSemanticScopeVisualDto;
import minic.uiapi.UiStageVisualDto;

import java.util.ArrayList;
import java.util.List;

/**
 * 根据 Semantic scope visual DTO 生成树形行模型。
 */
public final class MiniCSemanticScopeTreeModelFactory {
    /**
     * 创建作用域树行。
     *
     * @param visual 当前阶段 visual DTO
     * @return 树行
     */
    public List<MiniCSemanticScopeTreeLine> create(UiStageVisualDto visual) {
        if (visual.semanticRoot() == null) {
            return List.of();
        }
        ArrayList<MiniCSemanticScopeTreeLine> lines = new ArrayList<>();
        append(visual.semanticRoot(), 0, visual.semanticEdgesPointChildToParent(), lines);
        return List.copyOf(lines);
    }

    private boolean append(
            UiSemanticScopeVisualDto node,
            int depth,
            boolean childToParent,
            ArrayList<MiniCSemanticScopeTreeLine> lines
    ) {
        int currentIndex = lines.size();
        lines.add(new MiniCSemanticScopeTreeLine(
                node.label(),
                depth,
                node.symbols(),
                node.active(),
                node.active(),
                childToParent ? "child-to-parent" : "parent-to-child"
        ));
        boolean path = node.active();
        for (UiSemanticScopeVisualDto child : node.children()) {
            path |= append(child, depth + 1, childToParent, lines);
        }
        if (path && !lines.get(currentIndex).onActivePath()) {
            MiniCSemanticScopeTreeLine line = lines.get(currentIndex);
            lines.set(currentIndex, new MiniCSemanticScopeTreeLine(
                    line.label(),
                    line.depth(),
                    line.symbols(),
                    line.active(),
                    true,
                    line.arrowDirection()
            ));
        }
        return path;
    }
}

File: uiweb/src/visual/MiniCSemanticScopeTreeModelFactory.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiSemanticScopeVisualDto, UiStageVisualDto } from "../translation/uiapi";
import { MiniCSemanticScopeTreeLine } from "./MiniCSemanticScopeTreeLine";
import type { MiniCSemanticScopeArrowDirection } from "./MiniCSemanticScopeTreeLine";

export const miniCSemanticScopeTreeModelFactoryMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCSemanticScopeTreeModelFactory.java",
  webPath: "uiweb/src/visual/MiniCSemanticScopeTreeModelFactory.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCSemanticScopeTreeModelFactory",
  kind: "class",
  imports: ["minic.uiapi.UiSemanticScopeVisualDto", "minic.uiapi.UiStageVisualDto", "java.util.List"],
  fields: [],
  methods: [
    { name: "create", signature: "create(UiStageVisualDto visual)" },
    { name: "append", signature: "append(UiSemanticScopeVisualDto node, int depth, boolean childToParent, ArrayList<MiniCSemanticScopeTreeLine> lines)" },
  ],
} as const satisfies JavaMirrorFile;

export class MiniCSemanticScopeTreeModelFactory {
  static readonly mirror = miniCSemanticScopeTreeModelFactoryMirror;

  create(visual: UiStageVisualDto | null | undefined): readonly MiniCSemanticScopeTreeLine[] {
    if (!visual?.semanticRoot) {
      return [];
    }
    const lines: MiniCSemanticScopeTreeLine[] = [];
    this.append(visual.semanticRoot, 0, visual.semanticEdgesPointChildToParent, lines);
    return lines;
  }

  private append(
    node: UiSemanticScopeVisualDto,
    depth: number,
    childToParent: boolean,
    lines: MiniCSemanticScopeTreeLine[],
  ): boolean {
    const currentIndex = lines.length;
    const direction: MiniCSemanticScopeArrowDirection = childToParent ? "child-to-parent" : "parent-to-child";
    lines.push(new MiniCSemanticScopeTreeLine(node.label, depth, node.symbols, node.active, node.active, direction));
    let path = node.active;
    for (const child of node.children) {
      path = this.append(child, depth + 1, childToParent, lines) || path;
    }
    if (path && !lines[currentIndex].onActivePath) {
      const line = lines[currentIndex];
      lines[currentIndex] = new MiniCSemanticScopeTreeLine(
        line.label,
        line.depth,
        line.symbols,
        line.active,
        true,
        line.arrowDirection,
      );
    }
    return path;
  }
}

export default MiniCSemanticScopeTreeModelFactory;
